File: cpp/questions/numIslands.cpp
#include <bits/stdc++.h>
#include <vector>

using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rs = grid.size();
        int cs = grid[0].size();
        std::function<void(int, int)> step = [&](int r, int c) -> void {
            if (r<0 || r>=rs || c<0 || c>=cs || grid[r][c] == '0') {
                return;
            }
            grid[r][c] = '0';
            step(r-1, c);
            step(r+1, c);
            step(r, c-1);
            step(r, c+1);
        };
        int ans = 0;
        for (int r=0; r<rs; ++r) {
            for (int c=0; c<cs; ++c) {
                if (grid[r][c] == '1') {
                    ++ans;
                    step(r, c);
                }
            }
        }
        return ans; 
    }
};

File: python/questions/numIslands.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rl = len(grid)
        cl = len(grid[0])
        def step(r: int, c: int) -> None:
            nonlocal rl, cl
            if r<0 or r>=rl or c<0 or c>=cl or grid[r][c]=='0':
                return
            grid[r][c] = '0'
            step(r-1, c)
            step(r+1, c)
            step(r, c-1)
            step(r, c+1)
        ans = 0
        for i in range(rl):
            for j in range(cl):
                if grid[i][j] == '1':
                    ans += 1
                    step(i, j)
        return ans

